concordium-consensus/src/Concordium/GlobalState/Basic/BlockState/Accounts.hs

```diff
@@ -166,16 +167,16 @@ addressWouldClash addr Accounts{..} = AccountMap.addressWouldClashPure addr acco
 -- is the account index of the account is or was associated with, and @Nothing@
 -- otherwise.
 regIdExists :: ID.CredentialRegistrationID -> Accounts pv -> Maybe AccountIndex
-regIdExists rid Accounts{..} = rid `Map.lookup` accountRegIds
+regIdExists rid Accounts{..} = encode rid `Map.lookup` accountRegIds
 
 -- |Record an account registration ID as used on the account.
 recordRegId :: ID.CredentialRegistrationID -> AccountIndex -> Accounts pv -> Accounts pv
-recordRegId rid idx accs = accs { accountRegIds = Map.insert rid idx (accountRegIds accs) }
+recordRegId rid idx accs = accs { accountRegIds = Map.insert (encode rid) idx (accountRegIds accs) }
 
 -- |Record multiple registration ids as used. This implementation is marginally
 -- more efficient than repeatedly calling `recordRegId`.
 recordRegIds :: [(ID.CredentialRegistrationID, AccountIndex)] -> Accounts pv -> Accounts pv
-recordRegIds rids accs = accs { accountRegIds = Map.union (accountRegIds accs) (Map.fromAscList rids) }
+recordRegIds rids accs = accs { accountRegIds = Map.union (accountRegIds accs) (Map.fromAscList ((_1 %~ encode) <$> rids)) }
```

Comment:
I used it because there is a derived `Serialize`instance for `CredentialRegistrationID` (on which the `Ord` and `fromJSON` instances rely, and also `Serialize` instances for all types that have `CredentialRegistrationID` values as fields).

While I agree that serializing `CredentialRegistrationID` to `FixedByteString RegIdSize` will make the type more precise, an additional mental step is required to convince oneself that the two in-memory representations (one built by `encode`, another by the new `toRawCredId` function) are bitwise identical.

Would you agree that implementations of `encode` and `toRawCredId` shouldn't be independent from each other?

Reply:
Yes. It must hold that `encode = encode . toRawCredId`. So they are indeed not independent.

But this will trivially be the case since `toRawCredId` will be roughly `encode`, although it can be implemented a bit more efficiently than encoding to byte array and then converting to a fixed byte string.